### PR TITLE
Added GLBufferAttribute

### PR DESCRIPTION
**Pull request by raub**
If present, would delegate VBO creation to the user-supplied function. See more info on this in #11883. The default behavior is untouched.

### CODE REVIEW

**Comment by pailhead** [comment]
More hooks like this!!

**Comment by raub** [comment]
Now a new `DynamicBufferAttribute` class exists. Example: **webgl_buffergeometry_points_dynamicbufferattribute.html**.

`new DynamicBufferAttribute(...)` parameters:
* gl - renderer.context,
* buffer - an actual VBO,
* type - e.g. gl.FLOAT,
* itemSize - vector size, e.g. 3 for xyz position,
* count - vertex count.

**Comment by mrdoob** [comment]
Do you mind renaming it to `GLBufferAttribute`? `BufferAttribute` has a `dynamic` property (which sets `gl.DYNAMIC_DRAW) and I could see people getting confused.

**Comment by raub** [comment]
I see your point, but all of them (buffer types) are of GL kind... `GLBufferAttribute` is not a perfect name either. Could it be something else? Like
* `ManualBufferAttribute`
* `ExternalBufferAttribute`
* `SharedBufferAttribute`
* (other suggestions welcome)

I think prefix `GL` tells really nothing about what is special in this new attribute class, so it should be the last resort name of choise.

**Comment by raub** [comment]
@mrdoob, since no answer from you, I've decided to take some action and finally rename the class to `ExternalBufferAttribute`. Is this fine? What else is to be done before the merge could hapen?

**Comment by mrdoob** [comment]
I still prefer `GLBufferAttribute`. We're planning to also add `GLLine` for people that want to use raw `gl.LINE` instead of the new line with linewidth support. This name also hints the developer that no other renderers will be able to handle these classes.

Sorry I couldn't elaborate in time for r92, moving to r93.

**Comment by raub** [comment]
Ok, will do soon

**Comment by raub** [comment]
Done, `GLBufferAttribute`. Seems to work fine with the latest dev branch.

**Comment by WestLangley** [comment]
Nice demo. :-)

Can you please: add your new example to `examples/files.js`,  update the docs with your new class, and check your formatting. https://zz85.github.io/mrdoobapproves/

File: examples/webgl_buffergeometry_points_glbufferattribute.html
Comment (Mugen87) [inline examples/webgl_buffergeometry_points_glbufferattribute.html]: Maybe renaming the title to a more descriptive name: `three.js webgl - buffergeometry - custom VBOs`.

**Review by Mugen87 — CHANGES_REQUESTED** [review]

File: examples/webgl_buffergeometry_points_glbufferattribute.html
Comment (raub) [reply examples/webgl_buffergeometry_points_glbufferattribute.html]: Ok, see bbc5638

Comment (raub) [reply examples/webgl_buffergeometry_points_glbufferattribute.html]: Done

**Comment by raub** [comment]
@WestLangley
- Added the doc for `GLBufferAttribute` class. And listed it in the index.
- Added the example to the examples page.
- Could not find any more linting issues.

**Comment by raub** [comment]
@Mugen87 could you resolve the review? if there is something else to do, then please tell me

**Review by Mugen87 — APPROVED** [review]

File: src/core/BufferGeometry.js
Comment (mrdoob) [inline src/core/BufferGeometry.js]: How about instructing the user that the have to set `mesh.frustumCulled` to `false` instead?

Comment (mrdoob) [reply src/core/BufferGeometry.js]: We kind of do that already sometimes when using `SkinnedMesh` or when running gpu particle simulations.

Comment (WestLangley) [reply src/core/BufferGeometry.js]: >How about instructing the user that the have to set mesh.frustumCulled to false instead?

...or, alternatively, set the bounding sphere manually...

Comment (raub) [reply src/core/BufferGeometry.js]: Hmm, I didn't know about `mesh.frustumCulled`, this is very useful to me, thanks.

I think here it's better to check if the mesh is not culled or if it has some manual bbox/sphere. Because we don't pull the vertices from GPU to calc the bbox, but user might still need it. The warning is to be shown when both culling enabled and no bbox given. I'll try to implement this soon.

Comment (raub) [reply src/core/BufferGeometry.js]: I've simply added the warnings in f8f0b41. Now the example is set up to use either custom bounding sphere or no culling. I've checked that it shows the warning if I use neither of those in the example.

**Comment by raub** [comment]
@mrdoob Hey, is there anything else to be done before this gets merged?

Not only I wait for it because I hoped to switch my dependent module to use the main threejs package (rather then my fork), but also I'm getting manual merges here and there as my code is out of the common codebase.

**Comment by mrschofield** [comment]
Not to be a pain, but is there any chance this could be merged?

Is there any outstanding work that is required?

**Comment by mrdoob** [comment]
Okay, planning on merging this next week. Finally... 😅

File: docs/list.js
Comment (mrdoob) [inline docs/list.js:93]: This should be:

```js
"GLBufferAttribute": "api/en/core/GLBufferAttribute",
```

Comment (raub) [reply docs/list.js]: done

**Comment by Mugen87** [comment]
Can you please save a copy of `docs/api/en/core/GLBufferAttribute.html` at `docs/api/zh/core/GLBufferAttribute.html` and also add then Chinese entry it `docs/list.js`?

**Comment by Mugen87** [comment]
The TS declaration file for `GLBufferAttribute` should be something like:
```ts
export class GLBufferAttribute {

	constructor( gl: WebGLRenderingContext, buffer: WebGLBuffer, type: number, itemSize: number, count: number, normalized: boolean );

	buffer: WebGLBuffer;
	type: number;
	itemSize: number;
	elementSize: number;
	count: number;
	normalized: boolean;
	version: number;
	sizes: [ [ number, number ] ];

	readonly isGLBufferAttribute: true;

	set needsUpdate( value: boolean );

	setBuffer( buffer: WebGLBuffer ): this;
	setType( type: number ): this;
	setItemSize( itemSize: number ): this;
	setCount( count: number ): this;

}
```

File: src/core/GLBufferAttribute.js
Comment (Mugen87) [inline src/core/GLBufferAttribute.js:8]: Is this property supposed to be private? Is so I would remove it from the public API.

File: src/core/GLBufferAttribute.js
Comment (Mugen87) [inline src/core/GLBufferAttribute.js:32]: Is `elementSize` also intended to be private?

File: src/core/GLBufferAttribute.js
Comment (raub) [reply src/core/GLBufferAttribute.js:8]: The object stored in `this.sizes` is used by methods below. I also need `gl` to build this object. This is why I build it at the construction time, when `gl` is by definition available. Being a 'public' member is just a side-effect of the above. What would you suggest?

File: src/core/GLBufferAttribute.js
Comment (raub) [reply src/core/GLBufferAttribute.js:32]: It is intended to be public. The idea is as follows: you know the type of your data, you pass it, the `elementSize` becomes available, carrying the size-in-bytes value for you.

It is not necessary. We could let the app do all the calculations and leave here only what's necessary for three.js to handle the attribute.

Comment (raub) [reply src/core/GLBufferAttribute.js:32]: Ah wait, I found it here: src\renderers\webgl\WebGLAttributes.js
```
			buffers.set( attribute, {
				buffer: attribute.buffer,
				type: attribute.type,
				bytesPerElement: attribute.elementSize,
				version: attribute.version
			} );
```
so basically it is there for three.js to get the byte size, because there is no `Array` instance to do so (`bytesPerElement: array.BYTES_PER_ELEMENT,`).

**Comment by raub** [comment]
`sizes: [ [ number, number ] ];` I think should be `sizes: { [number]: number };` something. Unsure if this is a fine piece of TS syntax though. How can we validate it?

**Comment by Mugen87** [comment]
If you are unsure about this, use `any` for the moment^^.

File: src/core/GLBufferAttribute.js
Comment (Mugen87) [reply src/core/GLBufferAttribute.js:32]: Um, have you considered to let the user set this value? I mean `GLBufferAttribute` is obviously a low-level API not necessarily intended for beginners. Devs who use it should also know the correct bytes per element.

I would to put as less logic as possible into the ctor and let instead the user defined such information. In this way, you probably don't even need a WebGL rendering context in `GLBufferAttribute` which I find a bit strange anyway.

File: src/core/GLBufferAttribute.js
Comment (Mugen87) [reply src/core/GLBufferAttribute.js:8]: It seems you need `sizes ` only to determine `elementSize`, right? Check my comment below. I think it's better if the user defines `elementSize`.

File: src/core/GLBufferAttribute.js
Comment (raub) [reply src/core/GLBufferAttribute.js:32]: The logic behind this is that passing both `type` and `elementSize` to the constructor is redundant, because the `type` always implies the only possible `elementSize`. So it kind of creates a room for an error to occur when you ask the user to pass both. So I decided to calculate it on the ctor side.

But in general it would be fine either way. If the ctor takes `elementSize`, we get rid of all the logic and the contraversial `this.sizes` member. Also no need for `gl` then. Do you approve of the latter scenario?

Comment (Mugen87) [reply src/core/GLBufferAttribute.js:32]: I would say yes. Keeping `GLBufferAttribute` simple and let the app handle this bit sound like the better design for me. Especially since we can assume the user knows the relationship between `type` and `elementSize` if this class is going to be used.

The final decision is up to @mrdoob though 😉

Comment (raub) [reply src/core/GLBufferAttribute.js:32]: Any updates on that matter? @Mugen87 @mrdoob

Comment (Mugen87) [reply src/core/GLBufferAttribute.js:32]: Let's wait for @mrdoob's feedback.

**Comment by zlf2002171** [comment]
unsubscribe

2020-05-18 

zlf2002171 



发件人：Luis Blanco <notifications@github.com>
发送时间：2020-04-29 22:53
主题：Re: [mrdoob/three.js] Added GLBufferAttribute (#13196)
收件人："mrdoob/three.js"<three.js@noreply.github.com>
抄送："Subscribed"<subscribed@noreply.github.com>

@raub commented on this pull request.



In src/core/GLBufferAttribute.js:
> +
+		accum[ current[ 0 ] ] = current[ 1 ];
+		return accum;
+
+	}, {} );
+
+	if ( ! this.sizes[ type ] ) {
+
+		throw new TypeError( 'THREE.GLBufferAttribute: unsupported GL data type.' );
+
+	}
+
+	this.buffer = buffer;
+	this.type = type;
+	this.itemSize = itemSize;
+	this.elementSize = this.sizes[ type ];

Any updates on that matter? @Mugen87 @mrdoob

**Comment by pvaananen** [comment]
Sorry for bothering you but I'm really looking forward to this addition. So the current status is that the PR is waiting for @mrdoob's feedback?

Regarding that TypeScript question on Apr 27, may I suggest declaring an array of number pairs with `sizes:Array<[number, number]>` or `sizes:[number, number][]`.
```

**Comment by mrdoob** [comment]
Ah, too close to release now... Lets see if we can merge this at the beginning on the next dev cycle.

**Comment by Mugen87** [comment]
@mrdoob Before merging, can you please review https://github.com/mrdoob/three.js/pull/13196/files#r416037718? I'm not sure about the current public interface of `GLBufferAttribute`.

**Comment by raub** [comment]
1. Removed "gl" constructor param.
2. Removed "normalized" constructor param.
3. Added "elementSize" constructor param.
4. Removed "sizes" property.
5. Updated the TS class declaration.
6. Updated the docs.
7. Updated the example to fit new interface.
8. Updated the example logic for e2e compliance.
9. Debugged and fixed an issue when buffers don't change under VAO mode. (pls review!)
10. Generated a screenshot and passed e2e.

**Comment by Mugen87** [comment]
Thanks @raub! This looks much better to me now!

**Comment by mrdoob** [comment]
I think you can delete `webgl_buffergeometry_points_glbufferattribute.png`.

**Comment by mrdoob** [comment]
I'll merge and clean up.

**Comment by mrdoob** [comment]
Thanks!

File: src/renderers/webgl/WebGLBindingStates.js
Comment (Mugen87) [inline src/renderers/webgl/WebGLBindingStates.js:179]: Can you please explain why you have added this line? When removed, the respective demo does not seem to work correctly anymore. 

I don't think it's correct to always force a call of `setupVertexAttributes()` on each frame.

Comment (raub) [reply src/renderers/webgl/WebGLBindingStates.js:179]: I don't exactly understand what this part of the system is supposed to do. But my guess is:

- if there is no VAO, bind appropriate buffers
- if there is a VAO and the attributes were not changed, just switch to that VAO
- if attributes changed, update the VAO, i.e. switch to it and bind the updated buffers

That all sounds good in theory. But when I noticed my demo was broken, I started debugging and ended up here. So basically I set `needsUpdate` to my attribute and expect the engine to do the job of updating whatever was changed (the buffer). That was not the case here.

For some reason, `cachedAttribute` always contained an already-updated version of my attribute. So the VAO never got updated (hence the demo was broken). I double-checked this fact by adding some custom properties directly into the buffer, like `buffer.name=111`.

So I decided to opt-out from this optimization with my attribute type. I expect that line 179 effectively makes VAO unused for `GLBufferAttribute`. That is fine with me. Although if you understand what is going wrong and how to make it work with this optimization, that would also be great.

Comment (Mugen87) [reply src/renderers/webgl/WebGLBindingStates.js:179]: The problem is that the current implementation is bad for performance since a lot of unnecessary state changes are done. 

We definitely have to revisit this issue otherwise I would vote to revert `GLBufferAttribute` before the next release.

Comment (raub) [reply src/renderers/webgl/WebGLBindingStates.js:179]: The current implementation is vastly superior than having no implementation. And if I'm right, for `webgl1` people the further discussion makes no sense, since they always get exactly this level of performance. So there is a considerable group of people who either get this feature or not, regardless of VAO optimizations.

Now to optimizations. I would gladly fixed the VAO issue, given it was at all possible for you (or anybody) to explan why the **cached** attribute is always the updated attribute itself. And if it is the correct behavior, how it can be possibly fixed.

If you want, I might go on and report this behavior as a bug in a separate issue. Because now it is my understanding, that when I set `needsUpdate` on my attribute, the cache should contain the old version. If I have just updated the attribute, and the "cached" version does already contain what I have updated, **it is a bug in WebGLBindingStates** unless proven otherwise.

Comment (Mugen87) [reply src/renderers/webgl/WebGLBindingStates.js:179]: This has nothing to do with WebGL 1 or 2. VAO is available for most WebGL 1 users. Besides, even without VAO the buffer attribute bindings should only be done if necessary. Not if just the contents of a buffer changes.

I think the problem in your demo is this line:
```js
posAttr.buffer = ( posAttr.buffer === pos ) ? pos2 : pos;
```
You should exchange the entire attribute and not just the buffer. Or you keep the buffer the same and only change the contents.

Comment (Mugen87) [reply src/renderers/webgl/WebGLBindingStates.js:179]: This pattern works without the hack in `WebGLBindingStates`:
```js
var posAttr1 = new THREE.GLBufferAttribute( pos, gl.FLOAT, 3, 4, particles );
var posAttr2 = new THREE.GLBufferAttribute( pos2, gl.FLOAT, 3, 4, particles );
geometry.setAttribute( 'position', posAttr1 );

setInterval( function () {

    var attr = geometry.getAttribute( 'position' );

    geometry.setAttribute( 'position', ( attr === posAttr1 ) ? posAttr2 : posAttr1 );
 
}, 2000 );
```

**Pull request merged**
